File: cpp/steps/step07_joke_agent.cpp
// Agent that tells fresh dad jokes from API Ninjas.

#include <signalwire/agent/agent_base.hpp>
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <string>
#include <filesystem>

#include <httplib.h>
#include <nlohmann/json.hpp>

using namespace signalwire;
using json = nlohmann::json;

std::string check_ngrok() {
    try {
        httplib::Client cli("127.0.0.1", 4040);
        cli.set_connection_timeout(1);
        if (auto res = cli.Get("/api/tunnels")) {
            auto tunnels = json::parse(res->body);
            for (auto& t : tunnels.value("tunnels", json::array())) {
                if (t.value("proto", "") == "https") {
                    auto url = t.value("public_url", "");
                    setenv("SWML_PROXY_URL_BASE", url.c_str(), 1);
                    std::cout << "ngrok detected: " << url << "\n";
                    return url;
                }
            }
        }
    } catch (...) {}

    if (auto* env = std::getenv("SWML_PROXY_URL_BASE"); env && env[0]) {
        std::cout << "Using SWML_PROXY_URL_BASE from env: " << env << "\n";
        return env;
    }
    std::cout << "No ngrok tunnel detected and SWML_PROXY_URL_BASE not set\n";
    return "";
}

int main() {
    check_ngrok();

    agent::AgentBase agent("joke-agent");

    agent.add_language({"English", "en-US", "rime.spore"});

    agent.prompt_add_section(
        "Role",
        "You are a friendly assistant named Buddy. "
        "You love telling jokes and making people laugh. "
        "Keep your responses short since this is a phone call."
    );

    agent.prompt_add_section("Guidelines",
        "Follow these guidelines:", {
            "When someone asks for a joke, use the tell_joke function",
            "After telling a joke, pause for a reaction before offering another",
            "Be enthusiastic and have fun with it",
        }
    );

    agent.define_tool(
        "tell_joke",
        "Tell the caller a funny dad joke. Use this whenever someone "
        "asks for a joke, humor, or to be entertained.",
        {{"type", "object"}, {"properties", json::object()}},
        [](const json& args, const json& raw) -> swaig::FunctionResult {
            (void)args; (void)raw;

            auto* api_key = std::getenv("API_NINJAS_KEY");
            if (!api_key || !api_key[0]) {
                return swaig::FunctionResult(
                    "Sorry, I can't access my joke book right now. "
                    "My API key is missing.");
            }

            try {
                httplib::Client cli("https://api.api-ninjas.com");
                cli.set_connection_timeout(5);
                httplib::Headers headers = {{"X-Api-Key", api_key}};

                if (auto res = cli.Get("/v1/dadjokes", headers)) {
                    auto jokes = json::parse(res->body);
                    if (jokes.is_array() && !jokes.empty()) {
                        return swaig::FunctionResult(
                            "Here's a dad joke: " +
                            jokes[0].value("joke", ""));
                    }
                    return swaig::FunctionResult(
                        "I tried to find a joke but came up empty. "
                        "That's... kind of a joke itself?");
                }
            } catch (...) {}

            return swaig::FunctionResult(
                "Sorry, my joke service is taking a nap. "
                "Ask me again in a moment!");
        }
    );

    agent.set_post_prompt(
        "Summarize this conversation in 2-3 sentences. "
        "Note which jokes were told and how the caller reacted."
    );

    agent.on_summary([](const json& summary, const json& raw_data) {
        std::filesystem::create_directories("calls");
        auto call_id = raw_data.value("call_id",
            std::to_string(std::time(nullptr)));
        std::ofstream("calls/" + call_id + ".json") << raw_data.dump(2);
        std::cout << "Call summary saved: calls/" << call_id << ".json\n";
    });

    std::cout << "Starting joke-agent at http://0.0.0.0:3000/\n";
    agent.run();
}

File: java/steps/Step07JokeAgent.java
/**
 * Agent that tells fresh dad jokes from API Ninjas.
 *
 * Run:
 *   gradle run -PmainClass=JokeAgent
 *
 * Test:
 *   curl -s -u workshop:PASS http://localhost:3000/ | python3 -m json.tool
 */

import com.signalwire.agents.agent.AgentBase;
import com.signalwire.agents.swaig.FunctionResult;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;

public class JokeAgent {

    private static final Gson gson = new Gson();
    private static final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5))
            .build();

    static void checkNgrok() {
        try {
            var request = HttpRequest.newBuilder()
                    .uri(URI.create("http://127.0.0.1:4040/api/tunnels"))
                    .timeout(Duration.ofSeconds(1))
                    .GET()
                    .build();
            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            Map<String, Object> json = gson.fromJson(response.body(),
                    new TypeToken<Map<String, Object>>() {}.getType());
            @SuppressWarnings("unchecked")
            var tunnels = (List<Map<String, Object>>) json.get("tunnels");
            if (tunnels != null) {
                for (var t : tunnels) {
                    if ("https".equals(t.get("proto"))) {
                        var url = (String) t.get("public_url");
                        System.setProperty("SWML_PROXY_URL_BASE", url);
                        System.out.println("ngrok detected: " + url);
                        return;
                    }
                }
            }
        } catch (Exception e) {
            // ngrok not running
        }
        var current = System.getenv("SWML_PROXY_URL_BASE");
        if (current != null && !current.isEmpty()) {
            System.out.println("Using SWML_PROXY_URL_BASE from env: " + current);
        } else {
            System.out.println("No ngrok tunnel detected and SWML_PROXY_URL_BASE not set");
        }
    }

    /**
     * Call API Ninjas Dad Jokes endpoint and return the joke text.
     */
    static FunctionResult tellJoke(Map<String, Object> args, Map<String, Object> rawData) {
        var apiKey = System.getenv("API_NINJAS_KEY");
        if (apiKey == null || apiKey.isEmpty()) {
            return new FunctionResult(
                    "Sorry, I can't access my joke book right now. My API key is missing.");
        }

        try {
            var request = HttpRequest.newBuilder()
                    .uri(URI.create("https://api.api-ninjas.com/v1/dadjokes"))
                    .header("X-Api-Key", apiKey)
                    .timeout(Duration.ofSeconds(5))
                    .GET()
                    .build();
            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                return new FunctionResult(
                        "Sorry, my joke service is taking a nap. Ask me again in a moment!");
            }

            List<Map<String, Object>> jokes = gson.fromJson(response.body(),
                    new TypeToken<List<Map<String, Object>>>() {}.getType());
            if (jokes != null && !jokes.isEmpty()) {
                return new FunctionResult("Here's a dad joke: " + jokes.getFirst().get("joke"));
            }
            return new FunctionResult(
                    "I tried to find a joke but came up empty. That's... kind of a joke itself?");
        } catch (Exception e) {
            return new FunctionResult(
                    "Sorry, my joke service is taking a nap. Ask me again in a moment!");
        }
    }

    public static void main(String[] args) throws Exception {
        checkNgrok();

        var agent = AgentBase.builder()
                .name("joke-agent")
                .route("/")
                .port(3000)
                .build();

        agent.addLanguage("English", "en-US", "rime.spore");

        agent.promptAddSection("Role",
                "You are a friendly assistant named Buddy. "
                + "You love telling jokes and making people laugh. "
                + "Keep your responses short since this is a phone call.");

        agent.promptAddSection("Guidelines", "Follow these guidelines:", List.of(
                "When someone asks for a joke, use the tell_joke function",
                "After telling a joke, pause for a reaction before offering another",
                "Be enthusiastic and have fun with it"
        ));

        // Register joke function -- now calls live API
        agent.defineTool(
                "tell_joke",
                "Tell the caller a funny dad joke. Use this whenever someone asks for a joke, humor, or to be entertained.",
                Map.of("type", "object", "properties", Map.of()),
                JokeAgent::tellJoke
        );

        agent.setPostPrompt(
                "Summarize this conversation in 2-3 sentences. "
                + "Note which jokes were told and how the caller reacted.");

        agent.onSummary((summary, rawData) -> {
            try {
                Files.createDirectories(Path.of("calls"));
                var callId = rawData != null && rawData.containsKey("call_id")
                        ? (String) rawData.get("call_id")
                        : LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
                var filepath = Path.of("calls", callId + ".json");
                Files.writeString(filepath, gson.toJson(rawData));
                System.out.println("Call summary saved: " + filepath);
            } catch (IOException e) {
                System.err.println("Failed to save call summary: " + e.getMessage());
            }
        });

        System.out.println("Starting joke agent on port 3000...");
        agent.run();
    }
}
